server/core/src/config.rs

```diff
@@ -92,162 +112,154 @@ pub struct TlsConfiguration {
-            std::io::Error::new(std::io::ErrorKind::Other, e)
+    pub fn new<P: AsRef<Path>>(config_path: P) -> Result<Self, std::io::Error> {
+        // see if we can load it from the config file you asked for
+        eprintln!("📜 Using config file: {:?}", config_path.as_ref());
+        let mut f: File = File::open(config_path.as_ref()).inspect_err(|e| {
+            eprintln!("Unable to open config file [{:?}] 🥺", e);
+            let diag = kanidm_lib_file_permissions::diagnose_path(config_path.as_ref());
+            eprintln!("{}", diag);
         })?;
 
-        // check if the required fields are there
-        let mut config_failed = false;
-        if res.domain.is_none() {
-            eprintln!("❌ 'domain' field in server configuration is not set, server cannot start!");
-            config_failed = true;
-        }
+        let mut contents = String::new();
 
-        if res.origin.is_none() {
-            eprintln!("❌ 'origin' field in server configuration is not set, server cannot start!");
-            config_failed = true;
-        }
+        f.read_to_string(&mut contents).inspect_err(|e| {
+            eprintln!("unable to read contents {:?}", e);
+            let diag = kanidm_lib_file_permissions::diagnose_path(config_path.as_ref());
+            eprintln!("{}", diag);
+        })?;
 
-        if res.db_path.is_none() {
+        // if we *can* load the config we'll set config to that.
+        toml::from_str::<ServerConfigUntagged>(contents.as_str()).map_err(|err| {
             eprintln!(
-                "❌ 'db_path' field in server configuration is not set, server cannot start!"
+                "Unable to parse config from '{:?}': {:?}",
+                config_path.as_ref(),
+                err
             );
-            config_failed = true;
-        }
+            std::io::Error::new(std::io::ErrorKind::InvalidData, err)
+        })
+    }
+}
 
-        #[cfg(not(test))]
-        if res.tls_chain.is_none() {
-            eprintln!(
-                "❌ 'tls_chain' field in server configuration is not set, server cannot start!"
-            );
-            config_failed = true;
-        }
-        #[cfg(not(test))]
-        if res.tls_key.is_none() {
-            eprintln!(
-                "❌ 'tls_key' field in server configuration is not set, server cannot start!"
-            );
-            config_failed = true;
-        }
+#[derive(Debug, Deserialize, Default)]
+#[serde(deny_unknown_fields)]
+pub struct ServerConfigV2 {
+    domain: Option<String>,
+    origin: Option<String>,
+    db_path: Option<PathBuf>,
+    db_fs_type: Option<kanidm_proto::internal::FsType>,
+    tls_chain: Option<PathBuf>,
+    tls_key: Option<PathBuf>,
+    tls_client_ca: Option<PathBuf>,
+    bindaddress: Option<String>,
+    ldapbindaddress: Option<String>,
+    role: Option<ServerRole>,
+    log_level: Option<LogLevel>,
+    online_backup: Option<OnlineBackup>,
+    trust_x_forward_for: Option<bool>,
```

Comment:
Did we want to change this into an enum to preempt #3423 or are we trying to keep 100% compat with "v1"?

Reply:
The point of this change is just to enable versioning, then we will change this for V2 with #3423 - but there is no point adding the enum yet before we do that work, is there?

Reply:
Not really. But moving from "v1" to v2 is considered a breaking change no? But changing the bool to an enum variant is a breaking change too so it would need v3 of the config. Would make sense to set up v2 such that it takes #3423 into account so a second breaking v3 migration isn't required

Reply:
I think you're missing a valuable point here - v2 doesn't matter until we *release*. So once this change merges we can *keep changing it all we want* until the next release.

Reply:
Aight. If we committing to #3423 before next release then I am more than happy with that 🙂